platforms/core-execution/execution/README.md

```diff
@@ -38,96 +46,124 @@ Examples of units of work include:
+![](Execution%20Engine%20Schematic.drawio.svg)
 
-To leverage this, the execution engine allows units of work to define how their inputs should be **normalized**.
-Besides line-ending normalization for source files, Java compilation can also declare the compilation classpath to be normalized via _ABI extraction_.
-This means input hashes are calculated based on the extracted ABI of the classpath rather than raw file contents.
-This allows reuse of existing compilation results despite differences in dependencies, as long as those differences don't affect the ABI.
+> [!TIP]
+> There is a lot of useful detail in this schematic.
+> To open a zoomable version right-click on the image above and choose "Open Image in New Tab".
 
-Input normalization is a key feature that significantly increases the chance of reusing existing results, boosting execution performance.
+> [!TIP]
+> Use [draw.io](https://draw.io/) to make changes to the diagram. The linked SVG file can be opened directly in draw.io. Do not edit as raw SVG.
 
 ## Optimizations
 
-### Incremental Build (Up-to-Date Checks)
+### Up-to-Date Checks (aka Incremental Build)
 
-_TBD_
+> [!NOTE]
+> Incremental _build_ and incremental _work_ refer to different kinds of incrementality.
+> We should probably stop using incremental build as a term.
+
+This way we can compare the state of a unit of work before executing it to its state after its previous execution.
+If there are no changes to either its inputs or outputs, executing the work is not needed, and is thus skipped.
+
+* Input value changes are detected by comparing the `ValueSnapshot`s input value properties.
+
+* Input file changes are detected by comparing fingerprints of the input file properties.
+
+* Output changes are detected by comparing the snapshots of the output properties.
 
 ### The Build Cache
 
-_TBD_
+If the execution state of the work is out-of-date, we try to load outputs from the build cache.
+If successful, we delete and unpack results to the output location.
 
 ### Execution Modes
 
-Note that this differs from _incremental build,_ which involves skipping the entire execution of an up-to-date unit of work.
+When it comes to actually execution work, the execution engine supports two kinds of work: _mutable_ and _immutable._ _Mutable_ work can be executed _incrementally_ or _non-incrementally_ based on which of its inputs have changed compared to a previous execution.
 
-Fundamentally, the execution engine supports two kinds of work: _incremental_ and _non-incremental_.
+- **Immutable Work** is identified by its full set of inputs (i.e. all its inputs are **identity inputs**).
+  Examples are accessor generation and non-incremental artifact transforms.
 
-- **Non-Incremental Work**: Identified by its full set of inputs.
-  Changing any input effectively creates a new identity.
+- **Mutable Work**: Some of its inputs are **regular inputs** (while others can be **identity inputs**).
```

Comment:
The term "identity input" is not clearly defined. I guess it just means "other units of work" or something like that, but a bit of clarification would help.

Reply:
I tried to rephrase, is it better?